### PR TITLE
HPCC-13948 Fix divide by zero regression when retries exhausted

### PR DESCRIPTION
**Pull request by jakesmith**
Also fix an issue with loopCnt, which was causing it to be small
at start, but rose steadily.
This commit reverts the overall connect timeout to what it was
prior to HPCC-12128.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln @richardkchapman - please check

File: system/mp/mpcomm.cpp [hunk 1]
```diff
@@ -808,10 +808,9 @@ protected: friend class CMPPacketReader;
                                 EXCLOG(e, "MP: Failed to connect");
 #endif
                                 e->Release();
-                                if ((retrycount--==0)||(tm.timeout==MP_ASYNC_SEND))
+                                if ((--retrycount==0)||(tm.timeout==MP_ASYNC_SEND))
```
Comment (richardkchapman) [inline system/mp/mpcomm.cpp]: Not sure that's the best way to fix - I would have added a +1 on the code that divides by retrycount

**Comment by jakesmith** [comment]
Don't pull this PR, will change shortly

**Comment by jakesmith** [comment]
@mckellyln @richardkchapman - please check

File: system/mp/mpcomm.cpp [hunk 2]
```diff
@@ -756,8 +760,7 @@ protected: friend class CMPPacketReader;
                 // loop with short wait time and release CS to allow other side to proceed
                 StringBuffer epStr;
                 unsigned startMs = msTick();
-
-                unsigned loopCnt = ((CONNECT_READ_TIMEOUT / retrycount) / CONNECT_TIMEOUT_INTERVAL) + 1;
+                unsigned loopCnt = CONNECT_TIMEOUT_LOOPCNT;
```
Comment (mckellyln) [inline system/mp/mpcomm.cpp]: This makes the overall timeout for the connection a lot more that it was before.
Should we make it just based on the original value of retrycount.

Comment (jakesmith) [inline system/mp/mpcomm.cpp]: >This makes the overall timeout for the connection a lot more that it was before.

.. yes, because the retrycount was increased from 10 to 20. 

>Should we make it just based on the original value of retrycount.

```
unsigned loopCnt = ((CONNECT_READ_TIMEOUT / CONNECT_RETRYCOUNT) / CONNECT_TIMEOUT_INTERVAL) + 1;
```

Then each of these inner loops (i.e. each retrycount) is going to take [max] 10 secs...
Ok there's 20 retries vs original 10, but that's only [max] 200 secs overall.
NB: It used to be more, progressively, as retryCount decreased, but that was an accidental artefact.

The original [max] timeout was 1800s  (10 * 180s)


Why was retrycount upped? If it was reverted to 10 and each retry had this PR's fixed loop cnt (not based on retrycount) then it would be equivalent to what was there originally, wouldn't it?

Comment (mckellyln) [inline system/mp/mpcomm.cpp]: loopCnt is now 20x larger than before, right ?
I increased from 10 to 20 for more chances to check for other thread completing to speedup connection without wanting to affect overall timeout.

Comment (jakesmith) [inline system/mp/mpcomm.cpp]: >loopCnt is now 20x larger than before, right ?

Yes... but only at the start...
Your code made the overall time a lot small than it originally was..

It made:
1st retry pass = 10 secs (loopCnt = 10 * 1sec timeouts)
2nd retry pass = 10 secs (rounded down)
3rd retry pass = 11 secs
..
10th loop pass = 18 secs
...
19th loop pass = 180 secs

As I say, it climbed as retryCount fell.

The original pre-HPCC-12128 code, would take [up to] 180 secs per pass.

Comment (mckellyln) [inline system/mp/mpcomm.cpp]: More but shorter intervals so another thread gets a chance to aquire a CS was the goal for the smaller loopCnt and higher retrycount.  The error of using a variable retrycount and not a constant was bad.  But I don't think we need to make loopCnt 20x larger to fix that, just make it a constant equal to what is it is the first time through.

Comment (jakesmith) [inline system/mp/mpcomm.cpp]: The intervals are still shorter..

I assume that the rising time in each retry pass was unintentional (I think you confirmed)
I also assume that the overall descrese in time from the original code, which was 10 * 180 secs was unintentional.

Using a fixed loopCnt of 180 (= 180secs) restores the original overall time.. *except* the retriesCount was upped to 20 instead of 10 , and if there was no other reason , it should also be restored afaics.

**Comment by jakesmith** [comment]
@mckellyln - please check.

Changed so that total [max] timeout is restored to pre-HPCC-12128.
Mostly irrelevant anyway, because connect is passed a timeout which will be hit before it hits the retry limit.

**Comment by jakesmith** [comment]
@mckellyln - related, opened HPCC-13955 to allow the constants to be configurable and reduced.

**Comment by mckellyln** [comment]
looks good

File: system/mp/mpcomm.cpp [hunk 3]
```diff
@@ -756,8 +759,7 @@ protected: friend class CMPPacketReader;
                 // loop with short wait time and release CS to allow other side to proceed
                 StringBuffer epStr;
                 unsigned startMs = msTick();
-
-                unsigned loopCnt = ((CONNECT_READ_TIMEOUT / retrycount) / CONNECT_TIMEOUT_INTERVAL) + 1;
+                unsigned loopCnt = CONNECT_READ_TIMEOUT / CONNECT_TIMEOUT_INTERVAL;
```
Comment (richardkchapman) [inline system/mp/mpcomm.cpp]: Should there be a + 1 ?

Comment (mckellyln) [inline system/mp/mpcomm.cpp]: Yes I would suggest to add back the + 1.  Can I edit the branch to change ?

Comment (richardkchapman) [inline system/mp/mpcomm.cpp]: you can pull the branch and add another commit to it, but you will probably need to open a new PR unless you have write access to Jake's repo
